### DISCUSSION TITLE
Which is the Detekt public api?

### DISCUSSION THREAD

**Discussion by BraisGabin** [detekt/detekt#4117] [General]
I'm wondering which modules should be consider public api and which are internals. And then, for those that are public, what type of compatibility should we provide:

- none 
- api
- binary
- both

For example, I think that detekt-test and detekt-test-utils should be consider public. For example I use both in some custom rules: https://github.com/BraisGabin/detekt-junit-rules/blob/main/build.gradle.kts

But I think that we don't need to keep binary compatibility in those.

I think that this is important because we should simplify our public api for 2.0 because now it's a mess: For example, the cli module right now is kind of a public api but it shouldn't.

**Comment by picklebento** [comment]
Currently, we have `detekt-psi-utils`, `detekt-api` and `detekt-tooling` as public modules.
I checked the customized detekt repo in my company, which also depends on `detekt-test` directly and `detekt-test-utils` transitively.

I agree with you that we don't necessarily keep binary compatibility - It should be a judgment call to determine changes to those modules. On the other hand, we often draw a conclusion which maintains binary compatibility.
